pkg/provision/azure.go

```diff
@@ -0,0 +1,375 @@
+		"type": typeName,
+	}
+}
+
+func azureParameterValue(typeValue string) map[string]interface{} {
+	return map[string]interface{}{
+		"value": typeValue,
+	}
+}
+
+func getTemplateParameterDefinition() map[string]interface{} {
+	return map[string]interface{}{
+		"location":                  azureParameterType("string"),
+		"networkInterfaceName":      azureParameterType("string"),
+		"networkSecurityGroupName":  azureParameterType("string"),
+		"networkSecurityGroupRules": azureParameterType("array"),
+		"subnetName":                azureParameterType("string"),
+		"virtualNetworkName":        azureParameterType("string"),
+		"addressPrefixes":           azureParameterType("array"),
+		"subnets":                   azureParameterType("array"),
+		"publicIpAddressName":       azureParameterType("string"),
+		"virtualMachineName":        azureParameterType("string"),
+		"virtualMachineRG":          azureParameterType("string"),
+		"osDiskType":                azureParameterType("string"),
+		"virtualMachineSize":        azureParameterType("string"),
+		"adminUsername":             azureParameterType("string"),
+		"adminPassword":             azureParameterType("secureString"),
+		"customData":                azureParameterType("string"),
+	}
+}
+
+func getTemplateResourceVirtualMachine(host BasicHost) map[string]interface{} {
+	return map[string]interface{}{
+		"name":       "[parameters('virtualMachineName')]",
+		"type":       "Microsoft.Compute/virtualMachines",
+		"apiVersion": "2019-07-01",
+		"location":   "[parameters('location')]",
+		"dependsOn": []interface{}{
+			"[concat('Microsoft.Network/networkInterfaces/', parameters('networkInterfaceName'))]",
+		},
+		"properties": map[string]interface{}{
+			"hardwareProfile": map[string]interface{}{
+				"vmSize": "[parameters('virtualMachineSize')]",
+			},
+			"storageProfile": map[string]interface{}{
+				"osDisk": map[string]interface{}{
+					"createOption": "fromImage",
+					"managedDisk": map[string]interface{}{
+						"storageAccountType": "[parameters('osDiskType')]",
+					},
+				},
+				"imageReference": map[string]interface{}{
+					"publisher": host.Additional["imagePublisher"],
+					"offer":     host.Additional["imageOffer"],
+					"sku":       host.Additional["imageSku"],
+					"version":   host.Additional["imageVersion"],
+				},
+			},
+			"networkProfile": map[string]interface{}{
+				"networkInterfaces": []interface{}{
+					map[string]interface{}{
+						"id": "[resourceId('Microsoft.Network/networkInterfaces', parameters('networkInterfaceName'))]",
+					},
+				},
+			},
+			"osProfile": map[string]interface{}{
+				"computerName":  "[parameters('virtualMachineName')]",
+				"adminUsername": "[parameters('adminUsername')]",
+				"adminPassword": "[parameters('adminPassword')]",
+				"customData":    "[base64(parameters('customData'))]",
+			},
+		},
+	}
+}
+func getTemplateResourceNetworkInterface() map[string]interface{} {
+	return map[string]interface{}{
```

Comment:
This is so verbose compared to other APIs, is there not a way to template or generate this?

Please can you add a HTTP link to the doc where you found it / built it?

Reply:
There is another way, that we can put all these template and parameters into 2 JSON files. But the JSON files will be even more verbose because in JSON we cannot use go function to generate code with similar struct such as getSecurityRule in this PR. We also need to bundle the JSON files alone with the code when building binary.

This is the link regarding the JSON I'm talking about: https://github.com/Azure-Samples/azure-sdk-for-go-samples/tree/master/quickstarts/deploy-vm

So basically they are doing this:
```
func createDeployment() (deployment resources.DeploymentExtended, err error) {
	template, err := readJSON(templateFile)
	if err != nil {
		return
	}
	params, err := readJSON(parametersFile)
```

This was actually what I did in the first version, and then I figured it may be better to convert all the JSON into go struct / functions and I did it.

Reply:
We could keep a Go literal string instead too, that might be easier for maintenance. I don't want to read a file from disk for this.

Reply:
If we use literal string, that string will be very large text. And it would add 100+ lines of code in this azure.go because of no go code to generate duplicate/similar lines.

Another con of using literal string is that it actually makes the code logic a bit more complex. For example, right now I can just do:
```
"publisher": host.Additional["imagePublisher"],
```

With literal string, I will need to do something like
```
"publisher": "[parameter('imagePublisher')]",
```
and then add a new Azure parameter `imagePublisher` in `func getTemplateParameterDefinition` and then assign the value in the func `getParameter()`.

Basically from 1 place to 3 places to add a value.